### PR TITLE
Added basic equivalent to PowerShell modules

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
This is a WIP PR to start a conversation around making a replacement of `Ansible.ModuleUtils.Legacy.psm1` that covers more of what `basic.py` offers to Python modules. What I've done so far in the PR:

* Supports the majority of the argument spec handling that is present in basic.py
* ~~Better error handling in case of uncaught exceptions~~ Edit: done in https://github.com/ansible/ansible/pull/45334
* Return of invocation module args
* Support for no_log on individual options
* Integrated logging with the Windows Event log
* Expose a module object that can be passed to other module_utils so it can easily access parameters, module options
* 4 module `win_certificate_store`, `win_environment`, `win_ping`, and `win_uri` changed to use the newer format

Things still to do:

* ~~More validation with the argument spec value, easier to debug invalid spec structures~~ - done
* ~~Talk with the community around what they would like to see, more changes/features~~
* ~~Speed optimisations, currently it has a higher overhead than Legacy.psm1 due to there being more checks~~ - done moved to a C# class
* ~~Tests, tests, and more tests~~
* ~~Convert some more modules to use the new framework (maybe in another PR)~~
* ~~Documentation~~
* ~~Sanity tests to verify the argument spec matches the documentation~~ - will do in another PR

This is still an early draft and I'm not sure whether this will be the final form it will take. ~~I was originally planning on having this in C# but the Runspace/C# boundary with quitting the module made that difficult.~~ Figured this out and moved to a C# class.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
PowerShell

##### ANSIBLE VERSION
```
devel
```

### CODE REVIEW

File: lib/ansible/modules/windows/win_certificate_store.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_certificate_store.ps1]: Just a minor nitpick.
Do we want to use implicit comma-separated list syntax, or prefer explicit array syntax instead ?

I always get confused by the comma-separated syntax, but I am not sure what is standard practice in the PowerShell world.

Comment (dagwieers) [reply lib/ansible/modules/windows/win_certificate_store.ps1]: One reason for using the more explicit array definition is that it can be nested.
(*The* reason for me to prefer the `$( ... )` syntax over using backticks in shell)

File: lib/ansible/module_utils/powershell/Ansible.ModuleUtils.Basic.psm1
Comment (dagwieers) [inline lib/ansible/module_utils/powershell/Ansible.ModuleUtils.Basic.psm1]: I would prefer something that resembled  `exit_json` and `fail_json` more.

File: lib/ansible/modules/windows/win_certificate_store.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_certificate_store.ps1]: Aha, here is the check_mode support ;-)

File: lib/ansible/modules/windows/win_certificate_store.ps1
Comment (dagwieers) [reply lib/ansible/modules/windows/win_certificate_store.ps1]: Especially the fact that inside `$module = { ... }` there is a reference to `$Module` is very confusing.
I'd rather not put the module's essence inside an object.

**Comment by dagwieers** [comment]
For some reason I lost my review-comments. Here's a summary:
- I'd rather have the exit-functions names resemble `exit_json` and `fail_json`.
- I don't like the indirection (i.e. Invoke-AnsibleModule) of using $module with the code embedded in the object (for one it requires the whole module to be rewritten, but I also don't think it's as elegant as it originally was).
- Using $Module inside $module is very confusing IMO.
- I'd prefer if CheckMode support was defined together with the other module-defining information (not at the very bottom).

**Comment by jborean93** [comment]
@dagwieers thanks for the feedback, I'm currently trying a slightly different approach but here's some responses to your comments

> I'd rather have the exit-functions names resemble exit_json and fail_json.

There's 2 reasons for the current name (I'm not set on them anyway)

1. I can't name them `Exit-Json` and `Fail-Json` as they are already used in `Legacy.psm1` and would cause a name collision
2. `Fail` is not an [approved PS verb](https://docs.microsoft.com/en-us/powershell/developer/cmdlet/approved-verbs-for-windows-powershell-commands) this doesn't really matter as much

My new approach is to include them as part of the module object. I had troubles with this when using a `PSCustomObject` due to how the code is actually run but I'm hoping this new attempt with work out.

> I don't like the indirection (i.e. Invoke-AnsibleModule) of using $module with the code embedded in the object (for one it requires the whole module to be rewritten, but I also don't think it's as elegant as it originally was).

The biggest reasons for doing this is the ability for the module runner to catch uncaught exceptions and display a proper error message with a stacktrace when running under `-vvv`. For example here is an error message now

```
TASK [win_test2] ************************************************************************************************************************************************************************************************************************************************************************************************************
task path: /Users/jborean/dev/module-tester/adhoc.yml:6
Using module file /Users/jborean/dev/ansible/lib/ansible/modules/windows/win_test2.ps1
<SERVER2016.domain.local> ESTABLISH PSRP CONNECTION FOR USER: vagrant-domain@DOMAIN.LOCAL ON PORT 5986 TO SERVER2016.domain.local
<SERVER2016.domain.local> PSRP: EXEC (via pipeline wrapper)
fatal: [SERVER2016.domain.local]: FAILED! => {
    "changed": false,
    "module_stderr": "\r\nan uncaught exception\r\nAt line:18 char:5\r\n+     throw \"an uncaught exception\"\r\n+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : OperationStopped: (an uncaught exception:String) [], RuntimeException\r\n    + FullyQualifiedErrorId : an uncaught exception\r\n \r\n\r\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

compared to it running with `Invoke-AnsibleModule`

```
TASK [win_test] ***********************************************************************************************************************************************
task path: /Users/jborean/dev/module-tester/adhoc.yml:6
Using module file /Users/jborean/dev/ansible/lib/ansible/modules/windows/win_test.ps1
<SERVER2016.domain.local> ESTABLISH PSRP CONNECTION FOR USER: vagrant-domain@DOMAIN.LOCAL ON PORT 5986 TO SERVER2016.domain.local
<SERVER2016.domain.local> PSRP: EXEC (via pipeline wrapper)
The full traceback is:
an uncaught exception
At line:11 char:5
+     throw "an uncaught exception"
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (an uncaught exception:String) [], RuntimeException
    + FullyQualifiedErrorId : an uncaught exception

ScriptStackTrace:
at Test-Exception, <No file>: line 11
at Test-Func, <No file>: line 15
at <ScriptBlock>, <No file>: line 21
at Invoke-AnsibleModule, <No file>: line 1301
at <ScriptBlock>, <No file>: line 26
fatal: [SERVER2016.domain.local]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "another_option": null,
            "state": "present"
        }
    },
    "msg": "Unhandled exception while executing module (win_test): an uncaught exception"
}
```

The 2nd benefit that having the code being passed into a runner is that it can open up the way to having C# modules that utilise the same Module object. This is what I'm current working on right now.

With the comment around having to rewrite the module, I think that's going to be a given anyway. Currently we store the parameters and results locally whereas this new approach has them stored in the module object itself. When testing out the 2 modules it did require some work but that was mostly indenting the main code by 1 level and changing references of `$result` to `$Module.result`

> Using $Module inside $module is very confusing IMO.

Agreed, I originally had the input param as `$AnsibleModule` but shortened it for the testing without realising this collided with an existing variable.

> I'd prefer if CheckMode support was defined together with the other module-defining information (not at the very bottom).

Currently the `Invoke-AnsibleModule` takes in similar variables as `__init__()` in basic.py. We can always change how to call `Invoke-AnsibleModule` and use a splatting approach like so

```
#!powershell

#Requires -Module Ansible.ModuleUtils.Basic

$module_args = @{
    Arguments = $args
    ArgumentSpec = @{
        state = @{ type = "str"; default = "present"; choices = "absent", "present" }
        another_option = @{}
    }
    SupportsCheckMode = $true
}

$module = {
    param([PSTypeName("Ansible.Module")][PSCustomObject]$Module)

    Exit-AnsibleComplete -AnsibleModule $Module
}

Invoke-AnsibleModule -Module $module @module_args
```

**Comment by dagwieers** [comment]
> With the comment around having to rewrite the module, I think that's going to be a given anyway. 

Sure, but not just because everything needs to be re-indented. Obviously the new parameter-infrastructure requires those pieces to be adapted.

> Currently we store the parameters and results locally whereas this new approach has them stored in the module object itself. When testing out the 2 modules it did require some work but that was mostly indenting the main code by 1 level and changing references of $result to $Module.result

In python we don't do this though. We don't store the results in the AnsibleModule object and pass that object around rather than the required arguments. IMO that level of abstraction doesn't make it easier for contributors. It also doesn't make it easier to run pieces of code verbatim on the target machine for testing purposes.

**Comment by jborean93** [comment]
@dagwieers technically you don't need to indent it, it just needs the braces around it so it becomes a `ScriptBlock` but that's an ugly hack :). The alternative is to handle the runner on the leaf_exec side that calls the module currently. Will see what I can do to try and keep things similar to today.

> In python we don't do this though. We don't store the results in the AnsibleModule object and pass that object around rather than the required arguments

That's true, there's nothing stopping you from just passing the result object separately with this code. It's just now it is initialized and stored within the module object in this code. I could go either way with this but I find the conveniences of having the module object store the results dict convenient in some places.

**Comment by dagwieers** [comment]
> That's true, there's nothing stopping you from just passing the result object separately with this code. It's just now it is initialized and stored within the module object in this code. I could go either way with this but I find the conveniences of having the module object store the results dict convenient in some places.

I understand the advantages, in fact we are doing the same thing in our ACI modules (we have an ACIModule object storing everything, including AnsibleModule) and we only call its methods, which include our own exit_json and fail_json method. This makes every ACI module very similar in construct.
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/aci/aci_tenant.py

But the goal here was to make a similar experience to the python modules, and now we are surpassing what the python modules offer. That is one concern. The other concern is that we can't just drop code-pieces on the target node to test them, so the added abstraction layer makes it harder to test. (Which for ACI is not a concern as we cannot run code on the target anyway).

**Comment by jborean93** [comment]
@dagwieers 2nd round of changes have been made. I've addressed the following in this commit;

* Moved away from pure PowerShell code to a C# class for better performance, finer control over public properties as well as support for C# module in the future
* Because of the change above, I have now included `ExitJson` and `FailJson` as methods in this class removing the need for a separate cmdlet
* Added support for .NET Core (Nano server) with the JSON cmdlets if we ever move in that direction
* Removed the need for the module to be defined in a code block and keep things similar to how it works today
* Added the ability to define extra option types in a Scriptblock if a module need finer control over this
* Better validation of the argument spec hashtable before the params are parsed
* Better cleanup process when debugging in Visual Studio Code

The result obj is still part of the class and I need to talk about this a bit more with the wider community but I believe the recent changes address most of your concerns.

**Review by dagwieers — APPROVED** [review]
I like this a lot. And I hope we can move result into the module as a new standard.

**Comment by jborean93** [comment]
Things we want to change based on recent demo

* ~~Set the module invocation return values on result object straight after parsing the argspec/params~~
* Look further into the no_log validation, should we substrings that are a no_log value or just the full value itself as a match
* ~~Change the Log function to EventLog~~
* ~~Check whether WriteLine for a large buffer works with psrp, will it hang, should we change Console.WriteLine if a console exists or not~~
* ~~Add WriteLine for press enter to quite on Exit (non Runspace)~~
* Double check exit behaviour against various dev programs, ISE, VSCode, VS (full), Ansible, PS Console
* ~~Serializer can't serialize SecurityIdentifier on method invocation args~~
* ~~Add callable to type, remove ExtraTypes as this is not callable~~
* Look at type accelerators (maybe), will potentially opt out of this to as it adds extra code complexity and makes it harder to debug locally
* ~~Verify whether Set-StrictMode scope is based on the global or scoped to the function~~
* ~~Use generic IDictionary<String, object> for params~~

**Comment by jborean93** [comment]
Ok I think I'm at a stage where I am happy with what's in the PR. @dagwieers and @jhawkesworth it would be great if you could have a look over the newest changes and let me know how it looks from a developer perspective.

@acozine it would be great if you could look over the documentation changes and let me know what you think about the new additions.

@nitzmahone I believe I've incorporated the changes we discussed at Ansible Fest and have included tests. I did some further testing and can prove that `Ansible.Basic.cs` works on PSCore as long as https://github.com/ansible/ansible/pull/47643 has been merged.

ready_for_review

**Review by dagwieers — COMMENTED** [review]
Minor bikeshedding.
- I do still prefer `$module.Fail-Json()` and `$module.Exit-Json()` over `$module.FailJson()` and `$module.ExitJson()` (even prefer `$module.fail_json()`)
- Also I would prefer `$module.result` over `$module.Result` (if this becomes a thing in Python, it will be lowercase too)
- Same for `$module.check_mode` over `$module.CheckMode` (like in Python)
- And `$module.params.url` over `$module.Params.url`
- Also, we may need options to FailJson() and ExitJson() to set our own result value for those cases where we don't want to expose (incomplete) result data on failure (or exit).
- All messages returned to the user should be capitalized (it was like this before, but isn't the case anymore.
- The documentation lists parameter options (type, required, default, ..) together with the module options (required_if, mutual_exclusive, ...). Those should be separate.

Can't wait for this to be merged and start migrating modules over.

**Comment by jborean93** [comment]
Thanks @dagwieers for the feedback, here are some of my responses to your points, will have to let @nitzmahone weigh in.

> I do still prefer $module.Fail-Json() and $module.Exit-Json() over $module.FailJson() and $module.ExitJson() (even prefer $module.fail_json())

I'm not a fan of the PowerShell Verb-Noun format as this is a C# class. I tried to keep things to the [standard](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#capitalizing-compound-words-and-common-terms) for C# code but I'm not married to the concept. If anything I'm ok with moving towards a `snake_case` standard to match Python which will do the same for `Result`, `Params`, `CheckMode`, `Diff` and so on but that's something for us to agree upon

> Also, we may need options to FailJson() and ExitJson() to set our own result value for those cases where we don't want to expose (incomplete) result data on failure (or exit).

Is there a specific use case for this, I may be able to see it for `FailJson` but not `ExitJson`.

Edit: if you really want to, this is possible right now with `$module.Result = @{}` (or something like that) but that's not really that intuitive.

> All messages returned to the user should be capitalized (it was like this before, but isn't the case anymore.

I'm following the standard set in basic.py, most error messages created by the parser start as lowercase.

> The documentation lists parameter options (type, required, default, ..) together with the module options (required_if, mutual_exclusive, ...). Those should be separate.

The `required_if`, `mutually_exclusive`, and so on can also be set as a parameter option when `type=dict` or (`type=list` + `elements=dict`), maybe it's a matter of rewording the docs to better reflect that.

**Comment by nitzmahone** [comment]
I'm pretty much in agreement with @jborean93 about keeping the .NET classes using idiomatic .NET method naming. The ansible-facing bits should be snake-cased (eg, JSON returns), but the developer-facing bits should reflect the native language idioms where reasonable.

The only concern I have about adding a "return direct result" overload to `FailJson` is how we'd handle merging stuff like deprecation warnings into the result (eg, is it a "return exactly this", or a "still do a magic merge of other things"?). It seems like an attractive nuisance, given the propensity most people have not to read the docs; if Intellisense shows it there, people are going to tend to use it, even though it's only there for one specific case.

On the error messages, I don't have strong feelings either way, but the precedent's kinda been set by the Python side to start lowercase... If I were starting from scratch, I'd agree. If you want to float changing that everywhere, I could probably be convinced, but I don't think PS should be a snowflake in that regard.

**Review by nitzmahone — APPROVED** [review]
Impl is good to go, needs some docs wordsmithing which I'll volunteer for if @acozine isn't available...

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:296]: This bit should move up above lines 277-279 - otherwise we're telling people about referencing other C# utilities when they don't know C# utilities are possible . . .

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:194]: Are the parameters defined below specific to Windows modules?

File: changelogs/fragments/powershell_basic_util.yaml
Comment (acozine) [inline changelogs/fragments/powershell_basic_util.yaml]: How does adding C# module utils benefit the developer (or the user)? Is it faster? Easier to extend? Cleaner? More familiar? Also, are the new C# module utils replacing the old ones, or just augmenting them?

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:176]: Ah, this answers my question above - C# module utils are not replacing the PowerShell module utils, just adding to them. Would a developer choose only one, or could s/he use both?

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:197]: These are comprehensive and helpful. Adding examples would make them even better. Could we add examples, or link to an existing Windows module that uses all or most of these options?

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_windows.rst]: module_util not module_utils here

**Review by acozine — COMMENTED** [review]
@jborean93 nice work. I've added some questions and comments. When this PR is merged, I'd like to review the windows developer page with you in more detail  - it stayed unchanged during the Dev Guide update and could probably benefit from some attention.

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (jborean93) [reply docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:194]: Not 100%, they are shared with basic.py but the structure is slightly different.

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (jborean93) [reply docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:197]: Currently not all of these are used, I've just changed a few modules to use the new framework. The win_environment and win_uri module are linked in this doc as examples but as time goes on we will have more to reference to.

File: changelogs/fragments/powershell_basic_util.yaml
Comment (jborean93) [reply changelogs/fragments/powershell_basic_util.yaml]: Mostly that we already have PowerShell module_utils that embed C# code inside them. This makes things cleaner, more consistent in how they are loaded, and opens up the possibility of module utils if we do decide to add support for C# modules.

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (jborean93) [reply docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:176]: I'll add a bit around the choices between the 2.

File: docs/docsite/rst/dev_guide/developing_modules_general_windows.rst
Comment (jborean93) [reply docs/docsite/rst/dev_guide/developing_modules_general_windows.rst:296]: There's already a brief snippet around C# utils at https://github.com/ansible/ansible/pull/44705/files#diff-3caf171979d08135734172ec792f5820R176 which is above all this.

**Comment by jborean93** [comment]
Thanks for the review @acozine I've made a few changes. I agree that the guide needs an overhaul an am happy to help with that in a separate PR.

**Review by acozine — APPROVED** [review]
Docs component looks good for now. I'll revisit the Windows dev page with @jborean93 after this PR gets merged.

**Review by nitzmahone — APPROVED** [review]

**Comment by jborean93** [comment]
Thanks everyone for their reviews and help on this one.

**Pull request merged**
